packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -2,114 +2,73 @@ import 'uplot/dist/uPlot.min.css';
+import { usePlotConfig } from './hooks';
 import { PlotProps } from './types';
 
 // uPlot abstraction responsible for plot initialisation, setup and refresh
 // Receives a data frame that is x-axis aligned, as of https://github.com/leeoniya/uPlot/tree/master/docs#data-format
 // Exposes contexts for plugins registration and uPlot instance access
 export const UPlotChart: React.FC<PlotProps> = props => {
-  const theme = useTheme();
   const canvasRef = useRef<HTMLDivElement>(null);
-
-  // instance of uPlot, exposed via PlotContext
   const [plotInstance, setPlotInstance] = useState<uPlot>();
 
-  // Array with current plot data points, calculated when data frame is passed to a plot
-  // const [plotData, setPlotData] = useState<uPlot.AlignedData>();
-  // uPlot config
-  const [currentPlotConfig, setCurrentPlotConfig] = useState<uPlot.Options>();
+  // uPlot config API
+  const { currentConfig, addSeries, addAxis, addScale, registerPlugin } = usePlotConfig(
+    props.width,
+    props.height,
+    props.timeZone
+  );
 
-  // uPlot plugins API hook
-  const { arePluginsReady, plugins, registerPlugin } = usePlotPlugins();
+  const prevConfig = usePrevious(currentConfig);
 
   // Main function initialising uPlot. If final config is not settled it will do nothing
-  // Will destroy existing uPlot instance
-  const initPlot = useCallback(() => {
-    if (!currentPlotConfig || !canvasRef?.current) {
-      return;
+  const initPlot = () => {
+    if (!currentConfig || !canvasRef.current) {
+      return null;
     }
-
-    if (plotInstance) {
-      pluginLog('uPlot core', false, 'destroying existing instance due to reinitialisation');
-      plotInstance.destroy();
-    }
-
     const data = preparePlotData(props.data);
+    pluginLog('uPlot core', false, 'initialized with', data, currentConfig);
+    return new uPlot(currentConfig, data, canvasRef.current);
+  };
 
-    pluginLog('uPlot core', false, 'initialized with', data, currentPlotConfig);
-
-    setPlotInstance(new uPlot(currentPlotConfig, data, canvasRef.current));
-  }, [props, currentPlotConfig, arePluginsReady, canvasRef.current, plotInstance]);
-
-  const hasConfigChanged = useCallback(() => {
-    const config = buildPlotConfig(props, props.data, plugins, theme);
-    if (!currentPlotConfig) {
-      return false;
-    }
-
-    return shouldReinitialisePlot(currentPlotConfig, config);
-  }, [props, props.data, currentPlotConfig]);
-
-  // Initialise uPlot when config changes
-  useEffect(() => {
-    if (!currentPlotConfig) {
+  // Callback executed when there was no change in plot config
+  const updateData = useCallback(() => {
+    if (!plotInstance) {
       return;
     }
-    initPlot();
-  }, [currentPlotConfig]);
+    const data = preparePlotData(props.data);
+    pluginLog('uPlot core', false, 'updating plot data(throttled log!)');
+    // If config hasn't changed just update uPlot's data
+    plotInstance.setData(data);
+  }, [plotInstance, props.data]);
```

Comment:
nit: best practice is to spread props that are used as dependencies in hooks. I mean `setData` & `props.data` in this case

Reply:
My thinking here - setData never changes independently from plotInstance, that's why I used the entire instance as a dependency, to keep reasoning about this hook simpler. But if you do not agree that this is easier to grasp I'm happy to update :)

Reply:
That's ok for setData, but maybe props.data should be deconstructed and used instead? No biggie at all